### PR TITLE
ISLE: Allow callers to pass in storage for multi returns

### PR DESCRIPTION
**Pull request by fitzgen**
This allows callers to reuse storage and avoid allocations.

Internal constructors calling internal constructors won't reuse storage (at
least for now) but it is an option for the user code that is calling into ISLE.

Part of #7500

### CODE REVIEW

**Comment by fitzgen** [comment]
FWIW, no statistically significant changes to performance measured in sightglass. Still good to do this though, IMO.

File: cranelift/isle/isle/src/codegen.rs [hunk 1]
```diff
@@ -534,15 +626,16 @@ impl<'a> Codegen<'a> {
                     match ret_kind {
                         ReturnKind::Plain => write!(ctx.out, "return ")?,
                         ReturnKind::Option => write!(ctx.out, "return Some(")?,
-                        ReturnKind::Iterator => write!(ctx.out, "returns.push(")?,
+                        ReturnKind::Iterator => write!(ctx.out, "returns.extend(Some(")?,
```
Comment (elliottt) [inline cranelift/isle/isle/src/codegen.rs:629]: Is the choice of `Option` for the iterator here significant, or would `std::iter::once` also work?

**Review by elliottt — APPROVED** [review]
Nice!

File: cranelift/isle/isle/src/codegen.rs (hunk 1 above)
Comment (fitzgen) [reply cranelift/isle/isle/src/codegen.rs:629]: `std::iter::once` would also work, no strong feelings either way.

**Pull request merged**